Source/driver/timestep.cpp

```diff
@@ -223,3 +223,132 @@ Castro::estdt_temp_diffusion(void)
   return estdt_diff;
 }
 #endif
+
+bool
+Castro::check_timestep(Real dt)
+{
+    BL_PROFILE("Castro::check_timestep()");
+
+    MultiFab& S_new = get_new_data(State_Type);
+
+#ifdef REACTIONS
+    MultiFab& R_new = get_new_data(Reactions_Type);
+#endif
+
+    auto dx = geom.CellSizeArray();
+
+    ReduceOps<ReduceOpSum> reduce_op;
+    ReduceData<Real> reduce_data(reduce_op);
+    using ReduceTuple = typename decltype(reduce_data)::Type;
+
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+    for (MFIter mfi(S_new, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+
+        const Box& bx = mfi.tilebox();
+
+        auto U = S_new.array(mfi);
+#ifdef REACTIONS
+        auto R = R_new.array(mfi);
+#endif
+
+        reduce_op.eval(bx, reduce_data,
+        [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept -> ReduceTuple
+        {
+            Real zone_failed = 0.0_rt;
+
+            const Real derivative_floor = 1.e-50_rt;
+
+            Real dt_tmp = dt;
+
+            Real rhoInv = 1.0_rt / U(i,j,k,URHO);
+
+            // CFL hydrodynamic stability criterion
+
+            // If the timestep created a velocity v and sound speed at the new time
+            // such that (v+c) * dt / dx < CFL / change_max, where CFL is the user's
+            // chosen timestep constraint and change_max is the factor that determines
+            // how much the timestep can change during an advance, consider the advance
+            // to have failed. This prevents the timestep from shrinking too much, whereas
+            // in estdt change_max prevents the timestep from growing too much.
+
+            if (castro::do_hydro == 1) {
+
+                eos_t eos_state;
+
+                eos_state.rho = U(i,j,k,URHO );
+                eos_state.T   = U(i,j,k,UTEMP);
+                eos_state.e   = U(i,j,k,UEINT) * rhoInv;
+                for (int n = 0; n < NumSpec; ++n) {
+                    eos_state.xn[n] = U(i,j,k,UFS+n) * rhoInv;
+                }
+                for (int n = 0; n < NumAux; ++n) {
+                    eos_state.aux[n] = U(i,j,k,UFX+n) * rhoInv;
+                }
+
+                eos(eos_input_re, eos_state);
+
+                Real u = std::abs(U(i,j,k,UMX)) * rhoInv;
+                Real v = std::abs(U(i,j,k,UMY)) * rhoInv;
+                Real w = std::abs(U(i,j,k,UMZ)) * rhoInv;
+
+                Real c = eos_state.cs;
+
+                Real tau_CFL = amrex::min(AMREX_D_DECL(dx[0] / (c + u), dx[1] / (c + v), dx[2] / (c + w)));
```

Comment:
we should make this general for the true SDC stuff as well, with the different CFL definitions for different `time_integration_method`s

Reply:
I have an idea for how to do this. Ideally we'd call `estdt` on the new-time data and then just check if that timestep is within `change_max` of the old-time `dt` we started with. Then we don't need to have a separate routine, which is likely to diverge in ways like this. This would also solve the diffusion problem.